packages/unmock-core/src/service/state/validator.ts

```diff
@@ -134,124 +163,54 @@ const validStatesForStateWithoutCode = (
       relevantResponses[code] = filteredStateMedia;
     }
   }
-
   return Object.keys(relevantResponses).length > 0
     ? { responses: relevantResponses }
     : { error: err };
 };
-/**
- * Given a state and an operation, creates a copy of the requested state for each response that it matches.
- * First-level filtering is done via $code (status code) if it exists,
- * otherwise via matching the parameters set in `state`.
- * @param operation
- * @param state
- */
-export const getValidStatesForOperationWithState = (
-  operation: Operation,
-  state: IStateInputGenerator,
-  deref: Dereferencer,
-): {
-  responses: codeToMedia | undefined;
-  error: string | undefined;
-} => {
-  debugLog(
-    `getValidStatesForOperationWithState: Attempting to match and copy a ` +
-      `partial state that matches ${state} from ${operation}`,
-  );
-  const resps = operation.responses;
-  const code = state.top.$code;
-  // If $code is defined, we fetch the response even if no other state was set
-  // If $code is not found, use the default response - "the default MAY be used as a
-  //    default response object for all HTTP codes that are not covered individually by the specification"
-  // (see https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#responsesObject)
-  const { responses, error } =
-    code !== undefined
-      ? validStatesForStateWithCode(
-          resps[String(code) as codeType] || resps.default,
-          state,
-          code,
-          deref,
-        )
-      : // Otherwise, iterate over all status codes and find the ones matching the given state
-        validStatesForStateWithoutCode(resps, state, deref);
-  return { responses, error: error === undefined ? error : error.msg };
-};
 
 const getStateFromMedia = (
   contentRecord: Record<string, MediaType>,
   state: IStateInputGenerator,
   deref: Dereferencer,
 ): {
   responses: IResponsesFromContent;
-  error?: IMissingParam;
+  error?: IValidationError;
 } => {
   debugLog(
     `getStateFromMedia: Attempting to copy a partial state for ${state} from given media types ${contentRecord}`,
   );
-  const errors: IMissingParam[] = [];
+  let err: IValidationError | undefined;
   const relevantResponses: IResponsesFromContent = {};
   let success = false;
-  Object.keys(contentRecord).forEach((contentType: string) => {
+  for (const contentType of Object.keys(contentRecord)) {
```

Comment:
Personally I would some day like to get rid of these big for-loops that may be or may not have side effects, they are quite hard for me to follow but it doesn't matter now.

Reply:
It mostly looks big because of line breaks, debug logs and type checks, unfortunately. I don't like the use of `forEach` in these contexts.
There are no side effects either, a result is saved either to `error` or to a possible state.

When the day comes, I'd be happy to learn and see how you'd refactor this though to avoid a (for) loop :)

Reply:
Do you prefer the following, for example? (with setting `error` to `undefined` elsewhere -- in `validStatesForStateWithCode`)

```typescript
const getStateFromMedia = (
  contentRecord: Record<string, MediaType>,
  state: IStateInputGenerator,
  deref: Dereferencer,
): {
  responses: IResponsesFromContent;
  error?: IValidationError;
} => {
  debugLog(
    `getStateFromMedia: Attempting to copy a partial state for ${state} from given media types ${contentRecord}`,
  );
  return Object.entries(contentRecord).reduce(
    (
      acc: { responses: IResponsesFromContent; error?: IValidationError },
      [contentType, content],
    ) => {
      if (content == undefined || content.schema === undefined) {
        debugLog(`getStateFromMedia: No schema defined in ${contentType}`);
        acc.error = chooseBestMatchingError(
          {
            msg: `No schema defined in '${JSON.stringify(content)}'!`,
            nestedLevel: -1,
          },
          acc.error,
        );
        return acc;
      }
      const gen = state.gen(deref<Schema>(content.schema));
      if (gen.error !== undefined) {
        acc.error = chooseBestMatchingError(
          { msg: gen.error, nestedLevel: 0 },
          acc.error,
        );
        return acc;
      }
      debugLog(
        `getStateFromMedia: Copied matching state, verifying all state elements exist (not null)`,
      );

      debugLog(`getStateFromMedia: Spread state is valid for ${contentType}`);
      acc.responses[contentType] = gen.spreadState;
      return acc;
    },
    { responses: {} },
  );
};
```